feat: expose premium trial form via CLI

joobisb opened the pull request:
This PR closes https://github.com/coder/coder/issues/14856

joobisb committed:
feat: expose 30d trial form via CLI

joobisb commented on cli/login.go:
```diff
@@ -449,3 +553,75 @@ func openURL(inv *serpent.Invocation, urlToOpen string) error {
+}
+
+func promptCountry(inv *serpent.Invocation) (string, error) {
+	countries := []string{
```
this list is taken from [countries.tsx](https://github.com/coder/coder/blob/main/site/src/pages/SetupPage/countries.tsx) to have a select and filter option for this field, so the list will be maintained in two places, I'm open to options if there is a better way of handling this

johnstcn replied:
I would suggest making it a separate follow-up PR, but we could instead define countries in Go and have a separate script similar to `scripts/apitypings` to generate `countries.tsx` from `countries.go`.

joobisb replied:
sure, will address this in a separate PR, shall I create an issue to track this ?

johnstcn replied:
I went ahead and created one: https://github.com/coder/coder/issues/15074

johnstcn commented on cli/login.go:
```diff
@@ -149,6 +149,14 @@ func (r *RootCmd) login() *serpent.Command {
 		trial              bool
 		useTokenForSession bool
+
+		firstName   string
+		lastName    string
+		phoneNumber string
+		jobTitle    string
+		companyName string
+		country     string
+		developers  string
```
suggestion: let's instead move these declarations closer to L278

johnstcn commented on cli/login.go:
```diff
@@ -267,12 +275,66 @@ func (r *RootCmd) login() *serpent.Command {
 				}
 
+				if trial {
+					if firstName == "" {
```
suggestion: we can instead just leverage the zero value and assign directly to the fields of TrialInfo

e.g.

```
var trialInfo codersdk.CreateFirstUserTrialInfo
if trial {
  if trialInfo.FirstName == "" {
    ...
  }
  ...
}
```

johnstcn commented on cli/login.go:
```diff
@@ -398,6 +460,48 @@ func (r *RootCmd) login() *serpent.Command {
+			Env:         "CODER_FIRST_USER_FIRST_NAME",
+			Description: "Specifies the first name of the user.",
+			Value:       serpent.StringOf(&firstName),
+		},
+		{
+			Flag:        "first-user-last-name",
+			Env:         "CODER_FIRST_USER_LAST_NAME",
+			Description: "Specifies the last name of the user.",
+			Value:       serpent.StringOf(&lastName),
+		},
+		{
+			Flag:        "first-user-phone-number",
+			Env:         "CODER_FIRST_USER_PHONE_NUMBER",
+			Description: "Specifies the phone number of the user.",
+			Value:       serpent.StringOf(&phoneNumber),
+		},
+		{
+			Flag:        "first-user-job-title",
+			Env:         "CODER_FIRST_USER_JOB_TITLE",
+			Description: "Specifies the job title of the user.",
+			Value:       serpent.StringOf(&jobTitle),
+		},
+		{
+			Flag:        "first-user-company-name",
+			Env:         "CODER_FIRST_USER_COMPANY_NAME",
+			Description: "Specifies the company name of the user.",
+			Value:       serpent.StringOf(&companyName),
+		},
+		{
+			Flag:        "first-user-country",
+			Env:         "CODER_FIRST_USER_COUNTRY",
+			Description: "Specifies the country of the user.",
+			Value:       serpent.StringOf(&country),
+		},
+		{
+			Flag:        "first-user-developers",
+			Env:         "CODER_FIRST_USER_DEVELOPERS",
+			Description: "Specifies the number of developers.",
+			Value:       serpent.StringOf(&developers),
+		},
```
I'd rather not expose these as CLI arguments.

matifali replied:
I agree with Cian here.

joobisb replied:
got it, will remove these

johnstcn commented on cli/login_test.go:
```diff
@@ -142,6 +194,12 @@ func TestLogin(t *testing.T) {
 			"password", coderdtest.FirstUserParams.Password,
 			"password", coderdtest.FirstUserParams.Password, // confirm
 			"trial", "yes",
+			"firstName", coderdtest.TrialUserParams.FirstName,
+			"lastName", coderdtest.TrialUserParams.LastName,
+			"phoneNumber", coderdtest.TrialUserParams.PhoneNumber,
+			"jobTitle", coderdtest.TrialUserParams.JobTitle,
+			"companyName", coderdtest.TrialUserParams.CompanyName,
```
If trial is not set, then these parameters should not be required.

joobisb replied:
yeah that's right, but in this https://github.com/coder/coder/blob/10ccce1bd99210095895c27b21f04dda8d90cc30/cli/login_test.go#L196 trial is already set, so the params are required and I have added another test case `InitialUserTTYWithNoTrial` where trial is not set and params are not required
https://github.com/coder/coder/blob/10ccce1bd99210095895c27b21f04dda8d90cc30/cli/login_test.go#L128

johnstcn replied:
Oops, I must have missed that! Thanks for clarifying.

johnstcn reviewed (commented):
Thanks for your contribution! From my perspective, the trial information should always come directly from user input. I don't feel comfortable with allowing trials to be specified directly via CLI environment variables, although I admit there's very little stopping an enterprising user from feeding information in via `expect` or similar.

joobisb commented:
> Thanks for your contribution! From my perspective, the trial information should always come directly from user input. I don't feel comfortable with allowing trials to be specified directly via CLI environment variables, although I admit there's very little stopping an enterprising user from feeding information in via `expect` or similar.

@johnstcn should we remove `first-user-trial` as well from CLI env variables ? because if its `true` then other variables would also be required, or else we should make those optional

https://github.com/coder/coder/blob/10ccce1bd99210095895c27b21f04dda8d90cc30/cli/login.go#L453

johnstcn commented:
> @johnstcn should we remove `first-user-trial` as well from CLI env variables ? because if its `true` then other variables would also be required, or else we should make those optional
> 

Let's leave that as-is.

joobisb committed:
remove trial params as env

joobisb commented:
> > @johnstcn should we remove `first-user-trial` as well from CLI env variables ? because if its `true` then other variables would also be required, or else we should make those optional
> 
> Let's leave that as-is.

@johnstcn 
I've addressed the comments and I have checked through the codebase, the only issue I see is that of the use of `--first-user-trial` [here](https://github.com/coder/coder/blob/fac77f956ec28ad04c0c90750e381e69465d2ae1/.github/workflows/pr-deploy.yaml#L413), since the value is set this will prompt to enter the trial details causing the workflow to fail, one option is to set it as `--first-user-trial=false`, would that be ok?

https://github.com/coder/coder/blob/fac77f956ec28ad04c0c90750e381e69465d2ae1/.github/workflows/pr-deploy.yaml#L413

matifali commented:
@joobisb yes set that as false.

joobisb committed:
set first-user-trial to false in deploy-pr workflow

joobisb commented:
> @joobisb yes set that as false.

i've updated the PR

matifali commented on .github/workflows/pr-deploy.yaml:
```diff
@@ -410,7 +410,7 @@ jobs:
             --first-user-email pr${{ env.PR_NUMBER }}@coder.com \
             --first-user-password $password \
-            --first-user-trial \
+            --first-user-trial=false \
```
We are currently not making any use of a premium license in PR deployments. If needed, we can always upload one manually.

matifali commented on cli/login.go:
```diff
@@ -449,3 +496,75 @@ func openURL(inv *serpent.Invocation, urlToOpen string) error {
+		"Sudan", "Suriname", "Svalbard and Jan Mayen", "Swaziland", "Sweden", "Switzerland", "Syrian Arab Republic", "Taiwan, Province of China", "Tajikistan", "Tanzania, United Republic of",
+		"Thailand", "Timor-Leste", "Togo", "Tokelau", "Tonga", "Trinidad and Tobago", "Tunisia", "Turkey", "Turkmenistan", "Turks and Caicos Islands",
+		"Tuvalu", "Uganda", "Ukraine", "United Arab Emirates", "United Kingdom", "United States", "United States Minor Outlying Islands", "Uruguay", "Uzbekistan", "Vanuatu",
+		"Venezuela, Bolivarian Republic of", "Viet Nam", "Virgin Islands, British", "Virgin Islands, U.S.", "Wallis and Futuna", "Western Sahara", "Yemen", "Zambia", "Zimbabwe",
```
```suggestion
		"Venezuela, Bolivarian Republic of", "Vietnam", "Virgin Islands, British", "Virgin Islands, U.S.", "Wallis and Futuna", "Western Sahara", "Yemen", "Zambia", "Zimbabwe",
```

joobi-keyvalue replied:
this is fixed

joobisb replied:
this is fixed

joobisb committed:
fix typo

matifali commented:
Can you rebase on `main`?

johnstcn reviewed (commented):
One thing I've noticed is that in `develop.sh` we set `--first-user-trial=true`. This breaks the `develop.sh` workflow slightly as we then prompt for trial info. How about we instead just set `--first-user-trial=false`? In a separate PR we can modify the `develop.sh` script to upload a development license from a well-known location.

joobisb commented:
> One thing I've noticed is that in `develop.sh` we set `--first-user-trial=true`. This breaks the `develop.sh` workflow slightly as we then prompt for trial info. How about we instead just set `--first-user-trial=false`? In a separate PR we can modify the `develop.sh` script to upload a development license from a well-known location.

sure, I thought since it's used in the dev environment, we can always enter values into the prompt. But you are right, to make it more seamless I think we could set it to `false`.

joobisb committed:
Merge remote-tracking branch 'origin/main' into issue#14856

joobisb committed:
set first-user-trial to false in develop.sh

joobisb commented:
> > One thing I've noticed is that in `develop.sh` we set `--first-user-trial=true`. This breaks the `develop.sh` workflow slightly as we then prompt for trial info. How about we instead just set `--first-user-trial=false`? In a separate PR we can modify the `develop.sh` script to upload a development license from a well-known location.
> 
> sure, I thought since it's used in the dev environment, we can always enter values into the prompt. But you are right, to make it more seamless I think we could set it to `false`.

updated the PR

joobisb commented:
@matifali @johnstcn can we merge this ?

johnstcn commented:
@matifali it looks like the `e2e-premium` tests are failing. Can you try rebasing your PR on latest `main`?

joobisb committed:
Merge remote-tracking branch 'origin/main' into issue#14856

joobi-keyvalue commented:
> @matifali it looks like the `e2e-premium` tests are failing. @joobisb Can you try rebasing your PR on latest `main`?

@johnstcn done

joobisb commented:
> @matifali it looks like the `e2e-premium` tests are failing. @joobisb Can you try rebasing your PR on latest `main`?

@johnstcn done

joobisb commented:
@matifali can we merge this ?

matifali commented:
@sreya, we are blocked here as the secret is unavailable for forks. This is a required test so can not merge without it being passed.
